### PR TITLE
Certificate Transparency public API

### PR DESCRIPTION
**Pull request by RJPercival**
Reorganises the CT code so that some of the functions are public, whilst the remainder are mostly more private (e.g. data structures have mostly moved to ct_locl.h from ct_int.h). 

Tests now use the public API only.

Based on PR #632 - only the last commit is new (c680399).

### CODE REVIEW

**Comment by benlaurie** [comment]
+1 to c680399f92c88de5fab59af4cd97a00551a9dee4 only (the remainder being #632).

File: crypto/ct/ct_oct.c [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+                         *((c)++)=(unsigned char)(((l)>>16)&0xff), \
+                         *((c)++)=(unsigned char)(((l)>> 8)&0xff), \
+                         *((c)++)=(unsigned char)(((l)    )&0xff))
+
+int o2i_SCT_signature(SCT *sct, const unsigned char **in, size_t len)
+{
+    int ret = -1;
+    size_t siglen;
+    size_t len_remaining = len;
+    const unsigned char *p = *in;
```
Comment (ajmohan) [inline crypto/ct/ct_oct.c]: de-referrencing 'in' before its check for non-NULL

Comment (RJPercival) [inline crypto/ct/ct_oct.c]: Fixed.

**Comment by richsalz** [comment]
after you rebase, before pushing, please also make sure to update build.info file and/or test a --unified build :)  been there done that :(

**Comment by RJPercival** [comment]
Hi Rich, just to warn you, the commits you're pushing to fix things in the CT code are unnecessary. Everything you've addressed so far is fixed in the other PRs, so it'd be more efficient to just review them.

By the way, master is currently broken so I haven't pushed a rebased version of the PRs yet.

**Comment by richsalz** [comment]
Okay, so which PR's (in which order)?

**Comment by RJPercival** [comment]
They're in numeric order, starting with this one: #663, #664, #665, #666.

**Comment by richsalz** [comment]
Thanks.  But it's just build-fixes I think that went in so far.  What other things have you seen in master?

**Comment by RJPercival** [comment]
If you prefer, you can just review the last one though. They're only split up because Ben preferred reviewing them that way.

The current issue with master is that "HAVE_DLFCN_H" is no longer being defined by a Makefile, so I can't build successfully on Ubuntu. I believe that's already been discussed on the team mailing list and a fix is on its way, correct?

**Comment by ghedo** [comment]
@RJPercival having build fixes dependent on a huge changeset that is yet to be reviewed is not a good thing IMO

**Comment by ghedo** [comment]
...it just delays the merge of the build fixes for everyone else.

**Comment by RJPercival** [comment]
Sorry about that, the code originally came as one huge commit (if you think this changeset is big, you should have seen the original!) that I've teased apart and fixed. Since I've not been using the unified build much, I didn't notice that the first of the PRs was missing the updated build.info.

**Comment by levitte** [comment]
From what I can see, this looks sane from a build point of view.  With a little patience, we can see what Travis thinks about it.

**Comment by richsalz** [comment]
Correct, fix on the way. :)

**Comment by richsalz** [comment]
Address the parseable/valid issue and then I'll plus-one this.

**Comment by RJPercival** [comment]
I've now addressed that.

**Comment by richsalz** [comment]
+1
@benlaurie your turn :)

**Comment by benlaurie** [comment]
+1

**Comment by richsalz** [comment]
before I pull it ben, can you also do #752 ?

**Comment by richsalz** [comment]
pushed to master, thanks.

**Pull request closed without merging**